File: ios/TextAnalyzer.swift
import Foundation

// MARK: - Protocol for model path (plug in CoreML later)

protocol TextModelAnalyzing {
    func confidence(for text: String) -> Double
}

// No-op implementation used when no model is bundled
final class NoOpTextModelAnalyzing: TextModelAnalyzing {
    func confidence(for text: String) -> Double { 0.0 }
}

// MARK: - Error types

enum TextAnalyzerError: Error, LocalizedError {
    case invalidInput(String)
    case inferenceFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidInput(let msg):    return "INVALID_INPUT: \(msg)"
        case .inferenceFailed(let msg): return "INFERENCE_FAILED: \(msg)"
        }
    }
}

// MARK: - TextAnalyzer

final class TextAnalyzer {
    private let blocklistPatterns: [NSRegularExpression]
    private let modelBackend: TextModelAnalyzing

    init(blocklistURL: URL? = TextAnalyzer.defaultBlocklistURL(),
         modelBackend: TextModelAnalyzing = CoreMLTextModelAnalyzing.default) {
        self.modelBackend = modelBackend
        self.blocklistPatterns = TextAnalyzer.loadBlocklist(from: blocklistURL)
    }

    func analyze(
        input: String,
        threshold: Double,
        useBlocklist: Bool,
        useModel: Bool,
        extraTerms: [String]
    ) throws -> [String: Any] {
        guard !input.isEmpty else {
            throw TextAnalyzerError.invalidInput("input must be a non-empty string")
        }

        let start = ProcessInfo.processInfo.systemUptime
        let normalized = TextAnalyzer.normalize(input)

        var blocklistScore = 0.0
        if useBlocklist {
            let extraPatterns = extraTerms.compactMap { TextAnalyzer.makePattern(for: $0) }
            let allPatterns = blocklistPatterns + extraPatterns
            blocklistScore = TextAnalyzer.blocklistScore(normalized: normalized, patterns: allPatterns)
        }

        let modelScore = useModel ? modelBackend.confidence(for: normalized) : 0.0

        let confidence = max(blocklistScore, modelScore)
        let source: String
        if modelScore > blocklistScore && useModel {
            source = "tflite-text"
        } else {
            source = "blocklist"
        }

        let durationMs = Int((ProcessInfo.processInfo.systemUptime - start) * 1000)

        return [
            "isNSFW":     confidence >= threshold,
            "confidence": confidence,
            "threshold":  threshold,
            "source":     source,
            "durationMs": durationMs,
        ]
    }

    // MARK: - Private helpers

    static func normalize(_ text: String) -> String {
        var s = text.lowercased()
        let subs: [(Character, Character)] = [
            ("0", "o"), ("1", "i"), ("3", "e"),
            ("4", "a"), ("5", "s"), ("@", "a"), ("$", "s"),
        ]
        for (from, to) in subs {
            s = s.replacingOccurrences(of: String(from), with: String(to))
        }
        return s
    }

    static func blocklistScore(normalized: String, patterns: [NSRegularExpression]) -> Double {
        let range = NSRange(normalized.startIndex..., in: normalized)
        for pattern in patterns {
            if pattern.firstMatch(in: normalized, range: range) != nil {
                return 1.0
            }
        }
        return 0.0
    }

    static func makePattern(for term: String) -> NSRegularExpression? {
        let escaped = NSRegularExpression.escapedPattern(for: term.lowercased())
        return try? NSRegularExpression(pattern: "\\b\(escaped)\\b", options: .caseInsensitive)
    }

    static func loadBlocklist(from url: URL?) -> [NSRegularExpression] {
        guard let url = url,
              let content = try? String(contentsOf: url, encoding: .utf8) else {
            return []
        }
        return content
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && !$0.hasPrefix("#") }
            .compactMap { makePattern(for: $0) }
    }

    static func defaultBlocklistURL() -> URL? {
        // Look in the module bundle (bundled via podspec s.resources)
        let bundle = Bundle(for: TextAnalyzer.self)
        return bundle.url(forResource: "blocklist", withExtension: "txt")
    }
}

File: android/src/main/java/expo/modules/contentsafety/TextAnalyzer.kt
package expo.modules.contentsafety

import android.content.Context
import android.os.SystemClock
import android.util.Log

private const val TAG = "TextAnalyzer"

interface TextModelBackend {
    fun confidence(text: String): Double
}

class NoOpTextModelBackend : TextModelBackend {
    override fun confidence(text: String): Double = 0.0
}

class TextAnalyzer(
    private val context: Context,
    private val modelBackend: TextModelBackend = NoOpTextModelBackend(),
    private val blocklistLoader: (Context) -> List<Regex> = { loadBlocklist(it) }
) {
    private val blocklistPatterns: List<Regex> by lazy { blocklistLoader(context) }

    fun analyze(
        input: String,
        threshold: Double,
        useBlocklist: Boolean,
        useModel: Boolean,
        extraTerms: List<String>
    ): Map<String, Any> {
        if (input.isEmpty()) throw IllegalArgumentException("INVALID_INPUT: input must be a non-empty string")

        val start = SystemClock.elapsedRealtime()
        val normalized = normalize(input)

        val blocklistScore = if (useBlocklist) {
            val extraPatterns = extraTerms.map { termToRegex(it) }
            val allPatterns = blocklistPatterns + extraPatterns
            if (allPatterns.any { it.containsMatchIn(normalized) }) 1.0 else 0.0
        } else 0.0

        val modelScore = if (useModel) modelBackend.confidence(normalized) else 0.0

        val confidence = maxOf(blocklistScore, modelScore)
        val source = if (useModel && modelScore > blocklistScore) "tflite-text" else "blocklist"

        val durationMs = (SystemClock.elapsedRealtime() - start).toInt()

        return mapOf(
            "isNSFW"     to (confidence >= threshold),
            "confidence" to confidence,
            "threshold"  to threshold,
            "source"     to source,
            "durationMs" to durationMs,
        )
    }

    companion object {
        fun normalize(text: String): String {
            return text.lowercase()
                .replace('0', 'o')
                .replace('1', 'i')
                .replace('3', 'e')
                .replace('4', 'a')
                .replace('5', 's')
                .replace('@', 'a')
                .replace('$', 's')
        }

        fun termToRegex(term: String): Regex {
            val pattern = term.lowercase()
                .split(Regex("\\s+"))
                .joinToString("\\s+") { Regex.escape(it) }
            return Regex("\\b$pattern\\b", RegexOption.IGNORE_CASE)
        }

        fun loadBlocklist(context: Context): List<Regex> {
            return try {
                context.assets.open("blocklist.txt")
                    .bufferedReader()
                    .lineSequence()
                    .map { it.trim() }
                    .filter { it.isNotEmpty() && !it.startsWith("#") }
                    .map { termToRegex(it) }
                    .toList()
            } catch (e: Exception) {
                Log.w(TAG, "Failed to load blocklist.txt: ${e.message}")
                emptyList()
            }
        }
    }
}
